core/src/main/scala/org/http4s/Uri.scala

```diff
@@ -280,6 +279,110 @@ object Uri {
     }
   }
 
+  final class Path(
+      val segments: Chain[String],
+      val absolute: Boolean = false,
+      val endsWithSlash: Boolean = false)
```

Comment:
Seems to me you could get away with an empty segment at the beginning to represent absolute, and an empty segment at the end to represent endsWithSlash, and not need three variables.  But maybe that complicates other logic.

Reply:
Your idea of the ADT would move toward being able to represent absolute URIs at the type level, and could help solve the problem in #3210.  But it would need to carry forward to `Uri`.  Probably complicates things too much.  But ... maybe?